File: src/meshy/genParams.mjs
// genParams.mjs — central, single-source generation parameters for the Meshy
// text-to-3D pipeline.
//
// These literals are the ONLY place the preview/refine submit-body params live.
// Both the best-of-N CLI (scripts/meshy-generate.mjs) and the pipeline asset
// provider (src/pipeline/meshyAssetProvider.ts) import them, so the two paths
// submit byte-identical jobs and derive identical cache keys for the same prompt.
//
// Plain ESM (NO TypeScript syntax) so the CLI keeps running with no build step;
// the matching hand-written type declarations live alongside in genParams.d.ts.
//
// The disk cache key folds in paramSignature(mode) (see src/meshy/cache.mjs
// deriveKey), so changing ANY param below yields a NEW key and never silently
// serves stale bytes that were generated under different params.

// Preview submit params — tune the base mesh: remesh to a poly-bounded triangle
// mesh of at most target_polycount polygons.
export const PREVIEW_PARAMS = {
  should_remesh: true,
  target_polycount: 300000,
  topology: "triangle",
  ai_model: "meshy-6",
};

// Refine submit params — bake PBR texture onto the completed preview mesh so the
// pipeline returns a TEXTURED (not gray) GLB.
export const REFINE_PARAMS = {
  enable_pbr: true,
  remove_lighting: true,
  hd_texture: false,
  ai_model: "meshy-6",
};

// Canonical, key-sorted JSON of a value: recursively sorts object keys so the
// output is stable regardless of declaration order. This is what makes a param
// signature reproducible across processes and across the CLI / provider.
export function canonicalJson(value) {
  return JSON.stringify(sortKeysDeep(value));
}

function sortKeysDeep(value) {
  if (Array.isArray(value)) {
    return value.map(sortKeysDeep);
  }
  if (value !== null && typeof value === "object") {
    const sorted = {};
    for (const key of Object.keys(value).sort()) {
      sorted[key] = sortKeysDeep(value[key]);
    }
    return sorted;
  }
  return value;
}

// The output-affecting param signature for a mode, folded into the cache key by
// deriveKey. A refined asset is built ON TOP of a preview mesh, so the refine
// signature MUST include the preview params it was built on — otherwise changing
// a preview param (e.g. target_polycount) would not invalidate cached refined
// bytes that depend on it.
export function paramSignature(mode) {
  if (mode === "preview") {
    return canonicalJson(PREVIEW_PARAMS);
  }
  if (mode === "refine") {
    return canonicalJson({ preview: PREVIEW_PARAMS, refine: REFINE_PARAMS });
  }
  throw new Error(`paramSignature: unknown mode "${mode}" (expected "preview" or "refine")`);
}

File: backend/src/meshy/gen_params.py
"""Central, single-source generation parameters for the Meshy text-to-3D pipeline.

Mirrors src/meshy/genParams.mjs. These literals are the only place the
preview/refine submit-body params live. Changing any param yields a new cache
key so stale bytes are never served.
"""

from typing import Any

PreviewParams = dict[str, Any]
RefineParams = dict[str, Any]

PREVIEW_PARAMS: PreviewParams = {
    "should_remesh": True,
    "target_polycount": 300000,
    "topology": "triangle",
    "ai_model": "meshy-6",
}

REFINE_PARAMS: RefineParams = {
    "enable_pbr": True,
    "remove_lighting": True,
    "hd_texture": False,
    "ai_model": "meshy-6",
}


def _sort_keys_deep(value: Any) -> Any:
    """Recursively sort object keys for stable canonical JSON."""
    if isinstance(value, list):
        return [_sort_keys_deep(item) for item in value]
    if isinstance(value, dict):
        return {key: _sort_keys_deep(value[key]) for key in sorted(value.keys())}
    return value


def canonical_json(value: Any) -> str:
    """Return canonical, key-sorted JSON of a value."""
    import json

    return json.dumps(_sort_keys_deep(value), separators=(",", ":"))


def param_signature(mode: str) -> str:
    """Return the output-affecting param signature for a mode.

    The refine signature includes the preview params it was built on, so
    changing a preview param invalidates cached refined assets too.
    """
    if mode == "preview":
        return canonical_json(PREVIEW_PARAMS)
    if mode == "refine":
        return canonical_json({"preview": PREVIEW_PARAMS, "refine": REFINE_PARAMS})
    raise ValueError(f'param_signature: unknown mode "{mode}" (expected "preview" or "refine")')
